### PR TITLE
Full runtime config support

### LINKED ISSUE #84: Support Conform Style Configuration
The macro reads the configuration and sets it to the module attribute `@config`. Swoosh supports transforming environment variables, but that's only one way of deploying.

[bitwalker/conform](https://github.com/bitwalker/conform) is another model that overwrites the sys.config based on a configuration file deployed to the server. If the config was read at runtime, instead of compile time then the conform style of deployments could be supported as well.

### PR DESCRIPTION
**Pull request by princemaple**
Configs are loaded at runtime.
Module configs are kept, still having higher priority.

Configs now are read in the following order, the later ones take priority:

1. mix configs
```elixir
config :my_app, MyMailer, # the configs
```

2. module configs
```elixir
defmodule MyMailer do
  use Swoosh.Mailer, otp_app: :my_app, # other configs
end
```

3. function configs
```elixir
MyMailer.deliver(email, the_custom_configs)
```

4. system env
All the `{:system, "ENV_VAR"}` are finally loaded

close #84

### CODE REVIEW

File: lib/swoosh/mailer.ex [hunk 1]
```diff
@@ -82,14 +87,21 @@ defmodule Swoosh.Mailer do
...
   end
-  def deliver(_adapter, %Swoosh.Email{from: {_name, address}}, _config) when address in ["", nil] do
+  def deliver(%Swoosh.Email{from: {_name, address}}, _config)
+      when address in ["", nil] do
     {:error, :from_not_set}
   end
-  def deliver(adapter, %Swoosh.Email{} = email, config) do
-    config = Swoosh.Mailer.parse_runtime_config(config)
+  def deliver(%Swoosh.Email{} = email, {otp_app, mailer, module_config, config}) do
+    config =
```
Comment (stevedomin) [inline lib/swoosh/mailer.ex]: Probably worth extracting this into its own function, `parse_runtime_config` maybe?

File: lib/swoosh/mailer.ex [hunk 2]
```diff
@@ -82,14 +87,21 @@ defmodule Swoosh.Mailer do
...
     {:error, :from_not_set}
   end
-  def deliver(_adapter, %Swoosh.Email{from: {_name, address}}, _config) when address in ["", nil] do
+  def deliver(%Swoosh.Email{from: {_name, address}}, _config)
+      when address in ["", nil] do
     {:error, :from_not_set}
   end
-  def deliver(adapter, %Swoosh.Email{} = email, config) do
-    config = Swoosh.Mailer.parse_runtime_config(config)
+  def deliver(%Swoosh.Email{} = email, {otp_app, mailer, module_config, config}) do
```
Comment (stevedomin) [inline lib/swoosh/mailer.ex]: Let's make it very explicit that `module_config` is a compile-time config by renaming it to something like `compiled_mailer_config` or `compiled_module_config`.

File: lib/swoosh/mailer.ex [hunk 3]
```diff
@@ -98,26 +110,26 @@ defmodule Swoosh.Mailer do
   @doc """
   Parses the OTP configuration at compile time.
   """
-  def parse_config(mailer, opts) do
-    otp_app = Keyword.fetch!(opts, :otp_app)
+  def parse_static_config(mailer, module_config) do
+    otp_app = Keyword.fetch!(module_config, :otp_app)
     config = Application.get_env(otp_app, mailer, [])
-    adapter = opts[:adapter] || config[:adapter]
```
Comment (stevedomin) [inline lib/swoosh/mailer.ex]: I don't think we should do this here anymore (getting the adapter module and raising an error if it's not set), it's something that should happen at runtime, in `deliver/2` or whatever function we use for parsing/merging config.

Comment (stevedomin) [inline lib/swoosh/mailer.ex]: In fact we're doing half of this line 104. I'd extract that and the check/raise in a separate function as well

Comment (princemaple) [inline lib/swoosh/mailer.ex]: removed and added a test to ensure raise on KeyError with `~r/:adapter not found/`
I was trying to catch this on compile time still, but given now users can pass adapter at runtime, I guess it's no longer necessary.

**Comment by princemaple** [comment]
@stevedomin please take another look

- extracted the config parsing into a function, calling it `parse_config` because it's taking both compile time and run time into account
- the only thing being checked at compile time is that `otp_app` is set, `adapter` is no longer checked
- adapter is `fetch!`'ed at run time

**Comment by stevedomin** [comment]
Looks good!

**Comment by stevedomin** [comment]
@princemaple pretty interesting article about config in Elixir libraries from @michalmuskala: http://michal.muskala.eu/2017/07/30/configuring-elixir-libraries.html

**Pull request merged**